### ISSUE TITLE
beforeload causing POST failure

### ISSUE THREAD

**Issue by keenan35i** [apache/cordova-plugin-inappbrowser#359]
Hi i really was happy about this feature as i needed it for my app, however when i enable beforeload, it fails to send post data to the server on a forum submit. The page load goes through but does not send the post data to the server

`<form accept-charset="UTF-8" action="/search_results?mobile_app=true" id="login" method="post">
`    

```
activeBrowser.addEventListener('beforeload', function(e, callBack){
        console.log('before load event triggered', e, e.url);
        //prints www.someurl.com/search_results?mobile_app=true
        // so when it redirects you dont get the post data
        callBack(e.url);
      });
```

**Comment by janpio** [comment]
What platform? What OS version?

I am not sure I am following. 
- Can you explain your problem and what you are trying to do a bit better? 
- How did you open the InAppBrowser window showing that page with the form? 
- Are you trying to "hook" into the POST? 
- Without the listener everything is working, but with it the POST data gets lost?

**Comment by keenan35i** [comment]
ios 12, i am trying to implement cacheing with the beforeload event, however if i include the beforeload even on a simple passthrough like the example i posted, post data within a form gets lost, the server gets the hit but without any post data. Everything works ok when i dont have the event listener or beforelaod=no.

`var options = 'beforeload=yes,usewkwebview=yes,location=no,toolbar=no,allowInlineMediaPlayback=yes,hidden=yes,disableAnimation=yes,hidespinner=yes';

cordova.InAppBrowser.open(url, '_blank', options);
`

**Comment by janpio** [comment]
Could you create a super simple reproduction app that helps reproduce this issue? Start a new Cordova app with `cordova create`, add the minimal code to reproduce and understand this, put it on GitHub and post a link to the repository here. Thanks.

**Comment by keenan35i** [comment]
I dont have a server i can give you access to to make the server hit, i made a slightly more complete html/js structure maybe this helps.

```
<html>
<head>
  <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,user-scalable=0,viewport-fit=cover">
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
</head>
<body>
  <form accept-charset="UTF-8" action="/search_results?mobile_app=true" id="login" method="post">
    <input type="text" name="someuser">
    <button type="submit">Continue</button>
  </form>
  testApp.init();
</body>
</html>

```

```
var testApp = {
  init: function(){
    var options = options || 'beforeload=yes,usewkwebview=yes,location=no,toolbar=no,allowInlineMediaPlayback=yes,hidden=yes,disableAnimation=yes,hidespinner=yes';
    testApp.activeBrowser = cordova.InAppBrowser.open(url, '_blank', options);
    
    testApp.activeBrowser.addEventListener('beforeload', function(e, callBack){
      console.log('before load event triggered', e, e.url);
      callBack(e.url);
    });
  }
};
```

**Comment by janpio** [comment]
> the form hits the server at "/search_results?mobile_app=true"

POST or GET?

> i made a slightly more complete html/js structure maybe this helps.

Probably will, but having a ready made app one just has to check out and build/run would be even better - saves a few minutes of work for the maintainer.

**Comment by keenan35i** [comment]
post, its in the form attribute, method="post" in the example

**Comment by janpio** [comment]
That I see. But is the request that is missing the POST data really a POST?

**Comment by keenan35i** [comment]
yes it is a POST

**Comment by keenan35i** [comment]
@dpa99c you added this feature correct? Any insights or ideas on what it could be, maybe it needs an alternate to using navigationAction.request.URL; as that seems to be only communicating the url without getting the data sent to it?

**Comment by dpa99c** [comment]
This feature was originally added by @wvengen in https://github.com/apache/cordova-plugin-inappbrowser/pull/276, but I patched it to the iOS WKWebView implementation in https://github.com/apache/cordova-plugin-inappbrowser/pull/350.

Does it work with the UIWebView implementation (`usewkwebview=no`)?

**Comment by keenan35i** [comment]
Not sure, however i dont think my app will work properly in UIwebview, it was built in wkwebview in a fork of the inappbrowser, but i switched back to master after the release of wkwebview. ill give it a shot though and give some feed back soon, thanks.

**Comment by keenan35i** [comment]
@dpa99c , I tested with "usewkwebview=no" same issue.

**Comment by dpa99c** [comment]
@keenan35i OK, that implies the issue must've been present in the original UIWebView implementation in #276 and that I've translated the same problem across to the WKWebView implementation in #350.
 
@wvengen any ideas how this could be resolved in your original UIWebView implementation (i.e. preservation of POST data in the intercepted request)?

**Comment by wvengen** [comment]
That's a good point - the callback basically loads the URL again, indeed POST data would not be present - I'm afraid even the method would get lost. This might also be an issue for the referer (depending on how that is derived).

To fix this, I think the following would be needed:
- include method and POST data in the event
- pass these on to the callback (e.g. as extra parameters)

When these arguments are missing, the `GET` method would be assumed.
I hope this data is available in the relevant callbacks, it depends on platform support. If the method is available but POST data not, we might want to skip the callback for POST requests.

**Comment by dpa99c** [comment]
On iOS with UIWebView, it should be possible to use the `NSURLRequest` object passed to `shouldStartLoadWithRequest` to capture the HTTP method type and the POST data.
We should then be able to use that to make a new request.

With WKWebView, while we can extract the POST data and request type from `NSURLRequest *request` property of the `WKNavigationAction` object passed in to `decidePolicyForNavigationAction`, according to [this SO post](https://stackoverflow.com/q/26253133/777265), it's not possible to programmatically make a POST request using WKWebView.

On Android it seems there might be an issue with intercepting POST requests in the first place.
According to [this SO post](https://stackoverflow.com/a/11562391/777265), `shouldOverrideUrlLoading()` isn't called on POST requests and it's that method [we're currently using](https://github.com/apache/cordova-plugin-inappbrowser/blob/master/src/android/InAppBrowser.java#L1137) to intercept the request on Android when `beforeload` is specified.
And according to [this SO post](https://stackoverflow.com/q/13954049/777265), the same is true for `shouldInterceptRequest()`.
So there doesn't seem to be an easy way of arbitrarily intercepting POST requests on Android.

**Comment by keenan35i** [comment]
In regards to IOS (wkwebview), why not just pass the POST data back to a js function within the plugin js that can handle creating a secondary POST function after the logic determines which url to hit if any.

**Comment by dpa99c** [comment]
@keenan35i that would be a possible workaround for WKWebView POST requests, but it's a bit hacky.
If POST request prove too problematic (i.e. Android also), we may need to restrict `beforeload` to only GET requests.

In the case of both GET and POST, we'd ideally want to preserve any request headers also, particularly given that custom request headers is something that's being considered as a feature in #361.

**Comment by keenan35i** [comment]
Ok in that case, can we disable listening to POST requests and rename it to beforeGETLoad, adding beforePOSTLoad in the future as a different feature, at least i would be able to use this feature instead of it causing the app to break.

**Comment by wvengen** [comment]
What about: `beforeload=get`, `beforeload=post` (to be developed later), `beforeload=get+post` (combination of the previous; or `=both`). For backward compatibility, we could let `beforeload=yes` mean `beforeload=get`.
Do we need to consider other methods like `DELETE`, `PUT` - I suppose one would generally want to intercept these methods together with `POST`.

**Comment by janpio** [comment]
Better solution, if really necessary, would probably be to only handle `GET` requests inside the function that does this and document that it only works for GET requests, agree?

**Comment by dpa99c** [comment]
This should work for both iOS implementations. 

On Android, we have no way of intercepting the initial POST request anyway so it will proceed (along with its POST data) regardless of whether `beforeload` is set and the `beforeload` callback will not be invoked.

**Comment by dpa99c** [comment]
@wvengen I'm working on a PR to address this issue.

iOS is done, but on Android I'm finding that on testing the existing implementation on the `master` branch, [`shouldOverrideUrlLoading()`](https://developer.android.com/reference/android/webkit/WebViewClient#shouldOverrideUrlLoading(android.webkit.WebView,%20java.lang.String)) is never being called, regardless of whether it's a POST or GET request being made by the InappBrowser.

I'm debugging it in Android Studio and while the breakpoint on [`inAppWebView.loadUrl()`](https://github.com/apache/cordova-plugin-inappbrowser/blob/master/src/android/InAppBrowser.java#L990) is being reached, the one on [`shouldOverrideUrlLoading()`](https://github.com/apache/cordova-plugin-inappbrowser/blob/master/src/android/InAppBrowser.java#L1133) is never hit.

Any idea why this might be? 

I'm testing on a device running Android 8.1 and using [this test project](https://github.com/dpa99c/cordova-plugin-inappbrowser-test) using `cordova-android@7.1.4`.

Note that if I insert an override for [`shouldInterceptRequest()`](https://developer.android.com/reference/android/webkit/WebViewClient#shouldInterceptRequest(android.webkit.WebView,%20android.webkit.WebResourceRequest)), then put a breakpoint on this, it does get hit (for every resource being loaded).

**Comment by dpa99c** [comment]
OK, I've figured it out: on Android `shouldOverrideUrlLoading()` is not called on initial navigation (to the initial URL passed in with `inAppWebView.loadUrl()`) but only on subsequent navigation (e.g. click a hyperlink or form submit button in the IAB content page) whereas `shouldInterceptRequest()` is also called on initial navigation.

On iOS (both implementations), the relevant callbacks are always invoked for both initial and subsequent navigation.

**Comment by dpa99c** [comment]
I've created a PR to address the cause of this issue and to leave the future possibility open of supporting interception of POST requests using `beforeload`.

Please give my branch a test: 

    cordova plugin add https://github.com/dpa99c/cordova-plugin-themeablebrowser#GH-359-fix-beforeload-POST

I've also created a test app container which I used to validate this:

    https://github.com/dpa99c/cordova-plugin-inappbrowser-test

**Comment by keenan35i** [comment]
The link you provided is for themeable browser plugin not in app browser plugin?

**Comment by dpa99c** [comment]
Don't be misled by the name: the repo is called `cordova-plugin-themeablebrowser` because it's originally a fork of [initialxy/cordova-plugin-themeablebrowser](https://github.com/initialxy/cordova-plugin-themeablebrowser) which is a fork of [apache/cordova-plugin-inappbrowser](https://github.com/apache/cordova-plugin-inappbrowser).

However the branch `GH-359-fix-beforeload-POST` is derived from `apache/cordova-plugin-inappbrowser#master`.
There are other branches derived from `initialxy/cordova-plugin-themeablebrowser`.

Fun with git: you can have multiple remotes!

**Comment by keenan35i** [comment]
ah ok thanks, will test it out and let you know if it works for my app.

**Comment by keenan35i** [comment]
Just tested it seems to be working great, `POST` is ignored when `beforeload=get`.

**Comment by keenan35i** [comment]
@dpa99c do you happen to have a branch with both this fix and the postMessage API feature together? Seems to be taking a while for the PR's to be merged and im trying to finish building my cacheing tool which uses both of these features.

**Comment by dpa99c** [comment]
@keenan35i I've kept the changes associated with both PRs separate so as not to confuse the PR review process. You can always checkout both of my PR branches and merge them into your own combined fork for now.

But yeah, it would be good to get this merged to master, as well as #362
